workflow/controller/operator.go

```diff
@@ -2387,6 +2404,30 @@ func (woc *wfOperationCtx) initializeNode(nodeName string, nodeType wfv1.NodeTyp
 	return &node
 }
 
+// Update a node status with cache status
+func (woc *wfOperationCtx) updateAsCacheNode(node *wfv1.NodeStatus, memStat *wfv1.MemoizationStatus) *wfv1.NodeStatus {
+	node.MemoizationStatus = memStat
+
+	woc.wf.Status.Nodes[node.ID] = *node
+	woc.updated = true
+
+	return node
+}
+
+// Update a node status that has been cached and marked as finished
+func (woc *wfOperationCtx) updateAsCacheHitNode(node *wfv1.NodeStatus, outputs *wfv1.Outputs, memStat *wfv1.MemoizationStatus, message ...string) *wfv1.NodeStatus {
+	node.Phase = wfv1.NodeSucceeded
+	node.Outputs = outputs
+	node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
+	node.MemoizationStatus = memStat
+
+	woc.wf.Status.Nodes[node.ID] = *node
+	woc.log.Infof("%s node %v updated %s%s", node.Type, node.ID, node.Phase, message)
+	woc.updated = true
+
+	return node
+}
```

Comment:
nitpick: I wonder if we could call `updateAsCacheNode()` from here? It looks like all of the functionality of that is in here. What do you think?

Reply:
You're right. I just try to mimic relation btw `initializeCacheNode` and `initializeCacheHitNode` but I think `updateAsCacheNode` and `updateAsCacheHitNode` is not like that. As you said I can call `updateAsCacheNode` here to apply DRY. Thanks!

Reply:
Plus, I wanna remove the return values. I think these functions are all side-effects with updating node and no need to return something.